[Clang][Sema] Fix the lambda call expression inside of a type alias declaration

zyn0217 opened the pull request:
This patch attempts to fix the lambda call expression inside of a type alias declaration from two aspects:
1. Defer the lambda call expression building until after we have sufficient template arguments. This avoids the overeager (and often wrong) semantic checking before the type alias instantiation.
2. Properly obtain template arguments involving a template type alias for constraint checking.

It is unfortunate that a `TypeAliasTemplateDecl` (or a `TypeAliasDecl`) is never a `DeclContext`, nor does it have an associated specialization Decl from which we could collect these template arguments. Thus, I added a new CodeSynthesisContext to record template arguments for alias declarations.

Fixes https://github.com/llvm/llvm-project/issues/70601
Fixes https://github.com/llvm/llvm-project/issues/76674
Fixes https://github.com/llvm/llvm-project/issues/79555
Fixes https://github.com/llvm/llvm-project/issues/81145
Fixes https://github.com/llvm/llvm-project/issues/82104

Note that this doesn't involve the fix for https://github.com/llvm/llvm-project/issues/28461. That seems different, and I'd like to leave it as a follow-up.

zyn0217 committed:
The lambda call inside of a type alias

zyn0217 committed:
Format & Comments

zyn0217 commented:
Result for libc++ tests:
```
********************
Failed Tests (1):
  llvm-libc++-shared.cfg.in :: libcxx/headers_in_modulemap.sh.py


Testing Time: 1232.89s

Total Discovered Tests: 9673
  Unsupported      :  585 (6.05%)
  Passed           : 9062 (93.68%)
  Expectedly Failed:   25 (0.26%)
  Failed           :    1 (0.01%)
```

The failing test seems to be my local configuration issue and doesn't relate to this patch IIUC.

zyn0217 commented:
Also added @shafik for some insights on tests / other existing issues.

zyn0217 committed:
fixup

zyn0217 commented:
Although this patch doesn't fix https://github.com/llvm/llvm-project/issues/82104 yet, which is a crash caused by a mismatch in the template argument during the return type deduction, I still tried a temporary workaround (turning some switches off / on for getTemplateInstantiationArgs), and that proves we can move forward the fix with this patch.

zyn0217 commented:
Friendly ping. I'm looking for your feedback before fixing other similar bugs, e.g. #82104.

erichkeane reviewed (commented):
I don't have any concerns, but want @cor3ntin to take a look.  He's under the weather at the moment, so he might be a few days.

cor3ntin commented on clang/lib/Sema/SemaConcept.cpp:
```diff
@@ -614,10 +614,14 @@ bool Sema::SetupConstraintScope(
-             PrimaryTemplate->getInstantiatedFromMemberTemplate();
-         FromMemTempl;
-         FromMemTempl = FromMemTempl->getInstantiatedFromMemberTemplate()) {
+    // Note that we shall not collect instantiated parameters from
+    // 'intermediate' transformed function templates but the primary template
+    // for which we have built up the template arguments relative to. Otherwise,
+    // we may have mismatched template parameter depth!
```
```suggestion
    // We should only collect instantiated parameters from 
    // the primary template. Otherwise, we may have mismatched template parameter depth!
```

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -313,9 +313,75 @@ Response HandleRecordDecl(const CXXRecordDecl *Rec,
+  if (Rec->isLambda()) {
     if (const Decl *LCD = Rec->getLambdaContextDecl())
       return Response::ChangeDecl(LCD);
+    // Attempt to retrieve the template arguments for a using alias declaration.
```
```suggestion
    // Retrieve the template arguments for a using alias declaration.
```

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -313,9 +313,75 @@ Response HandleRecordDecl(const CXXRecordDecl *Rec,
+    // Attempt to retrieve the template arguments for a using alias declaration.
+    // This is necessary for constraint checking, since we always keep
+    // constraints relative to the primary template.
+    if (ForConstraintInstantiation && !SemaRef.CodeSynthesisContexts.empty()) {
```
```suggestion
    if (ForConstraintInstantiation) {
```

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -313,9 +313,75 @@ Response HandleRecordDecl(const CXXRecordDecl *Rec,
+        // unfortunate that we only have the mappings of call operators rather
+        // than lambda classes.
+        while (true) {
+          auto *FTD = dyn_cast_if_present<FunctionTemplateDecl>(
+              LambdaCallOperator->getDescribedTemplate());
+          if (FTD && FTD->getInstantiatedFromMemberTemplate()) {
+            LambdaCallOperator =
+                FTD->getInstantiatedFromMemberTemplate()->getTemplatedDecl();
+          } else if (auto *Prev = cast<CXXMethodDecl>(LambdaCallOperator)
+                                      ->getInstantiatedFromMemberFunction())
+            LambdaCallOperator = Prev;
+          else
+            break;
+        }
```
```suggestion
          if (auto *FTD = dyn_cast_if_present<FunctionTemplateDecl>(
              LambdaCallOperator->getDescribedTemplate(); FTD && FTD->getInstantiatedFromMemberTemplate()) {
            LambdaCallOperator =
                FTD->getInstantiatedFromMemberTemplate()->getTemplatedDecl();
          } else if (auto *Prev = cast<CXXMethodDecl>(LambdaCallOperator)
                                      ->getInstantiatedFromMemberFunction())
            LambdaCallOperator = Prev;
          else
            break;
        }
```

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -313,9 +313,75 @@ Response HandleRecordDecl(const CXXRecordDecl *Rec,
+          else
+            break;
+        }
+        // Same applies for type alias Decl. We perform this to obtain the
```
> Same applies for type alias Decl

I'm not sure I understand this comment. What does "Same" refer to? Can you rephrase?

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -313,9 +313,75 @@ Response HandleRecordDecl(const CXXRecordDecl *Rec,
+            break;
+        }
+        // Same applies for type alias Decl. We perform this to obtain the
+        // "canonical" template parameter depths.
```
```suggestion
        // canonical template parameter depth.
```

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -313,9 +313,75 @@ Response HandleRecordDecl(const CXXRecordDecl *Rec,
+        // "canonical" template parameter depths.
+        while (TATD->getInstantiatedFromMemberTemplate())
+          TATD = TATD->getInstantiatedFromMemberTemplate();
+        // Tell if we're currently inside of a lambda expression that is
```
```suggestion
        // Check if we are currently inside of a lambda expression that is
```

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -313,9 +313,75 @@ Response HandleRecordDecl(const CXXRecordDecl *Rec,
+        //  1. The lambda expression and the using alias declaration share the
+        //  same declaration context.
+        //  2. They have the same template depth.
+        // Then we assume the template arguments from the using alias
+        // declaration are essential for constraint instantiation. We have to do
```
```suggestion
        // Then we assume the template arguments of the using alias
        // declaration are necessary for constraint instantiation. We have to do
```

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -313,9 +313,75 @@ Response HandleRecordDecl(const CXXRecordDecl *Rec,
+          Result.addOuterTemplateArguments(CurrentTATD,
+                                           CSC.template_arguments(),
+                                           /*Final=*/false);
+          // Visit the parent of the current type alias declaration rather than
+          // the lambda thereof. We have the following case:
```
```suggestion
          // Visit the parent of the current type alias declaration rather than
          // the lambda thereof. 
          E.g, in the following example:
```

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -313,9 +313,75 @@ Response HandleRecordDecl(const CXXRecordDecl *Rec,
+          // void foo() {
+          //   S::T var;
+          // }
+          // The instantiated lambda expression (which we're visiting at 'var')
+          // has a function DeclContext 'foo' rather than the Record DeclContext
+          // S. This seems to be an oversight that we may want to set a Sema
+          // Context from the CXXScopeSpec before substituting into T to me.
```
@erichkeane Opinion?

cor3ntin commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -614,6 +682,15 @@ Sema::InstantiatingTemplate::InstantiatingTemplate(
+    : InstantiatingTemplate(
+          SemaRef, Sema::CodeSynthesisContext::TypeAliasTemplateInstantiation,
+          PointOfInstantiation, InstantiationRange, /*Entity=*/Template,
+          nullptr, TemplateArgs) {}
```
Can you add a comment for the `nullptr`?

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -13905,6 +13911,31 @@ TreeTransform<Derived>::TransformLambdaExpr(LambdaExpr *E) {
   SavedContext.pop();
 
+  // Recompute the dependency of the lambda so that we can defer the lambda call
+  // construction until after we have sufficient template arguments. For
```
```suggestion
  // construction until after we have all the necessary template arguments. For
```

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -13905,6 +13911,31 @@ TreeTransform<Derived>::TransformLambdaExpr(LambdaExpr *E) {
 
+  // Recompute the dependency of the lambda so that we can defer the lambda call
+  // construction until after we have sufficient template arguments. For
+  // example, template <class> struct S {
+  //   template <class U>
```
```suggestion
  // example, given
  // template <class> struct S {
  //   template <class U>
```

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -13905,6 +13911,31 @@ TreeTransform<Derived>::TransformLambdaExpr(LambdaExpr *E) {
+  //   using T = S<int>::Type<float>;
+  //             ^~~~~~
+  // }
+  // We would end up here from instantiating the S<int> as we're ensuring the
+  // completeness. That would make us transform the lambda call expression
+  // despite the fact that we don't see the argument for U yet. We have a
```
```suggestion
  // We would end up here from instantiating S<int> when ensuring its
  // completeness. That would make us transform the lambda call expression
  // despite the fact that we don't have a corresponding argument for U yet. We have a
```

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -13905,6 +13911,31 @@ TreeTransform<Derived>::TransformLambdaExpr(LambdaExpr *E) {
+  // despite the fact that we don't see the argument for U yet. We have a
+  // mechanism that circumvents the semantic checking if the CallExpr is
+  // dependent. We can harness that by recomputing the lambda dependency from
+  // the instantiation arguments. I'm putting it here rather than the above
+  // since we can see transformed lambda parameters in case that they're
+  // useful for calculation.
```
I am not sure i understand this comment, can you rephrase?

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -13905,6 +13911,31 @@ TreeTransform<Derived>::TransformLambdaExpr(LambdaExpr *E) {
+  // the instantiation arguments. I'm putting it here rather than the above
+  // since we can see transformed lambda parameters in case that they're
+  // useful for calculation.
+  DependencyKind = getDerived().ComputeLambdaDependency(&LSICopy);
+  Class->setLambdaDependencyKind(DependencyKind);
+  // Clean up the type cache created previously. Then, we re-create a type for
+  // such Decl with the new DependencyKind.
+  Class->setTypeForDecl(nullptr);
+  getSema().Context.getTypeDeclType(Class);
+
```
This is not great, but I am not sure how we can improve. 
I think it's worth a fixme
@erichkeane

cor3ntin commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -267,6 +267,11 @@ Bug Fixes to C++ Support
   (`#782154 <https://github.com/llvm/llvm-project/issues/82154>`_`)
+- Clang now supports direct lambda calls inside of a type alias template declarations.
+  This addresses (`#70601 <https://github.com/llvm/llvm-project/issues/70601>`_),
+  (`#76674 <https://github.com/llvm/llvm-project/issues/76674>`_),
```
```suggestion
  (`#GH76674),
```
We added that syntax recently, if you want

zyn0217 replied:
Yeah, the patch needs a rebase later :)

zyn0217 committed:
Slightly refactor & Fix GH82104

zyn0217 committed:
Mention GH82104 in the release note

zyn0217 committed:
Rephrase & Format

Endilll reviewed (commented):
`Sema.h` changes look good to me.

zyn0217 commented:
So, I have:
1. rebased the patch on top of the recent Sema refactors.
2. incorporated the fix for https://github.com/llvm/llvm-project/issues/82104, which is a slight change on `HandleFunction`.
3. separated the `CodeSynthesisContext` iteration logic into three helper functions i.e. `getPrimaryTemplateOfGenericLambda`, `getEnclosingTypeAliasTemplateDecl` and `isLambdaEnclosedByTypeAliasDecl` - I hope this looks clearer.
4. clarified some comments.

In terms of the lambda dependencies, I'm not entirely sure if we can accept the current approach. In practice, it is possible to avoid such a hack at the moment, since we don't actually rely on the lambda substitution but rather the template arguments.

Do you think it will be useful in the future that we need something more e.g. the substituted parameters, to compute the dependency? @cor3ntin

jcsxky commented:
Probably I haven't understand these code deeply and I have some confusions on this patch. I wonder has this patch solved essential issue, or it just fixed the crash. I think some `static_assert` statements are required to demonstrate it does deduce the type of template parameters correctly.

zyn0217 committed:
Validate deduced types

zyn0217 committed:
Merge branch 'main' into lambda-call-inside-of-type-aliases

zyn0217 commented:
> I think some `static_assert` statements are required to demonstrate it does deduce the type of template parameters correctly.

Added. Although the patch doesn't touch that type deduction part and thus they are theoretically not affected.

zyn0217 commented:
Friendly ping @cor3ntin and @erichkeane if you're back in the offices.

erichkeane commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -80,6 +80,81 @@ struct Response {
+// than lambda classes.
+const FunctionDecl *
+getPrimaryTemplateOfGenericLambda(const FunctionDecl *LambdaCallOperator) {
+  while (true) {
```
What cases do we have where this takes more than 1 step?  I also wonder if we'd be better off finding the primary `CXXRecordDecl`, then picking it up from there?  We could use `getTemplateInstantiationPattern` then `getDescribedClassTemplate` I think?

zyn0217 replied:
(Sorry for replying this late, I just got around to this PR.)

> What cases do we have where this takes more than 1 step?
The lambda itself might be defined within a template, and I presume transforming that template introduces multilevel mappings.

> I also wonder if we'd be better off finding the primary `CXXRecordDecl`, then picking it up from there? We could use `getTemplateInstantiationPattern` then `getDescribedClassTemplate` I think?

If we examine `TemplateInstantiator::transformedLocalDecl`, we'd see there's only handling for `CXXMethodDecls` of adding such mappings that we can extract from the Decls themselves. I think it is possible that we can use the primary `CXXRecordDecl` approach, although looking for that Decl might involve handling `LocalInstantiationScopes`, which IMO is not super straightforward.

erichkeane reviewed (commented):
1 thought, else I don't see anything.  @cor3ntin is coming back in a few days, so I'd like him to do a final once over.

cor3ntin reviewed (approved):
I think this looks reasonable.
Thanks for the patch!

shafik commented:
ping @cor3ntin

zyn0217 commented:
Thank you folks for the insightful review! I'm merging this PR now.

jcsxky commented:
Still crash on
```cpp
template<auto F> constexpr auto x = F();
template<class> constexpr int a() { return 1; }

template <class>
struct A {
    using Func = decltype(
    []<class T>(T) {
        return x<[] constexpr { return a<T>(); }>;
        // return x<[] constexpr { return b(); }>;
    }.template operator()<unsigned long long>('2')
    );
};
A<short>::Func y;
```

zyn0217 commented:
> Still crash on
> 
> ```c++
> template<auto F> constexpr auto x = F();
> template<class> constexpr int a() { return 1; }
> 
> template <class>
> struct A {
>     using Func = decltype(
>     []<class T>(T) {
>         return x<[] constexpr { return a<T>(); }>;
>         // return x<[] constexpr { return b(); }>;
>     }.template operator()<unsigned long long>('2')
>     );
> };
> A<short>::Func y;
> ```

That is a distinct case: Func does *not* form a `TypeAliasTemplateDecl`, and thus it is *not* supposed to be covered by this PR. Please submit a new issue for it.

jcsxky commented:
> > Still crash on
> > ```c++
> > template<auto F> constexpr auto x = F();
> > template<class> constexpr int a() { return 1; }
> > 
> > template <class>
> > struct A {
> >     using Func = decltype(
> >     []<class T>(T) {
> >         return x<[] constexpr { return a<T>(); }>;
> >         // return x<[] constexpr { return b(); }>;
> >     }.template operator()<unsigned long long>('2')
> >     );
> > };
> > A<short>::Func y;
> > ```
> 
> That is a distinct case: Func does _not_ form a `TypeAliasTemplateDecl`, and thus it is _not_ supposed to be covered by this PR. Please submit a new issue for it.

Consider this one,
```cpp
template<auto F> constexpr auto x = F();
template<class> constexpr int a() { return 1; }

template <class>
struct A {
    template<typename U>
    using Func = decltype(
    []<class T>(T) {
        return x<[] constexpr { return a<T>(); }>;
    }.template operator()<unsigned long long>('2')
    );
};
A<short>::Func<int> y;
```
`Func` is a `TypeAliasTemplateDecl` and crashes as well.

zyn0217 commented:
@jcsxky At the glance of the stacktrace, I suspect we probably need some mechanism of deferral codegen while instantiating the enclosing struct. Can you please file a separate issue?

shafik commented:
This crash looks linked to this PR: https://github.com/llvm/llvm-project/issues/136432
